### PR TITLE
[utils] Add more options for testing a function

### PR DESCRIPTION
**Pull request by gergondet**
- Configure the count of random configurations
- Allow to pass specific configurations

### CODE REVIEW

File: include/tvm/utils/checkFunction.h [hunk 1]
```diff
@@ -5,26 +5,55 @@
...
 {
 public:
+  /** A specific configuration to check
+   * - \a value will be stored in the function's variable value
+   * - \a velocity will be stored in the function's variable's derivative value
+   * - \a acceleration will be stored in the function's variable's derivative's
+   *   derivative value
+   * - \a samples if any of value/velocity/acceleration is null then \a samples
+   *   random configuration are generated for those (if it makes sense for this test)
+   */
```
Comment (aescande) [inline include/tvm/utils/checkFunction.h:41]: I would add that the order in the vectors corresponds to the order given by the `VariableVector` of the function tested.

File: include/tvm/utils/checkFunction.h [hunk 2]
```diff
@@ -5,26 +5,55 @@
...
+    size_t samples = 1;
+  };
+
   CheckOptions() : step(1e-7), prec(1e-6), verbose(true) {}
   CheckOptions(double s, double p, bool v) : step(s), prec(p), verbose(v) {}
   double step;
   double prec;
   bool verbose;
+  size_t samples = 1;
+  std::vector<CheckConfiguration> configs = {};
```
Comment (aescande) [inline include/tvm/utils/checkFunction.h:56]: A small inconvenience: if we want to just test a given configuration, we need to remember to set `samples` to `0`.
Maybe we could have `samples` to 0 by default and add a bit of logic to ensure at least one sample is run if `samples==0` and `configs.size()==0`.

File: src/utils/checkFunction.cpp [hunk 3]
```diff
@@ -141,42 +237,105 @@ bool TVM_DLLAPI checkVelocity(FunctionPtr f, CheckOptions opt)
...
+    return getFromConfigOrRandom<&CheckConfiguration::velocity>(prefix, c, vel);
+  };
+  for(const auto & c : opt.configs)
+  {
+    if(c.value && c.velocity)
+    {
+      if(getValue(c) && getVelocity(c))
+      {
+        total_fixed++;
+        failed_fixed += checkVelocity(uf, opt, val, vel);
```
Comment (aescande) [inline src/utils/checkFunction.cpp:285]: I think there's a missing `!`.

File: src/utils/checkFunction.cpp [hunk 4]
```diff
@@ -141,42 +237,105 @@ bool TVM_DLLAPI checkVelocity(FunctionPtr f, CheckOptions opt)
...
+      }
+    }
+    else
+    {
+      for(size_t i = 0; i < c.samples; ++i)
+      {
+        if(getValue(c) && getVelocity(c))
+        {
+          total_fixed++;
+          failed_fixed += checkVelocity(uf, opt, val, vel);
```
Comment (aescande) [inline src/utils/checkFunction.cpp:295]: Idem

**Review by aescande — COMMENTED** [review]
Thanks for these improvements. I just have a couple of remarks and found two missing `!`.

File: src/utils/checkFunction.cpp [hunk 5]
```diff
@@ -215,13 +374,120 @@ bool TVM_DLLAPI checkNormalAcceleration(FunctionPtr f, CheckOptions opt)
...
+    {
+      std::cout << failed_fixed << " configurations failed out of " << total_fixed << "\n";
+    }
+  }
+  return (failed + failed_fixed) == 0;
+}
+
+} // namespace
+
+bool TVM_DLLAPI checkNormalAcceleration(FunctionPtr f, CheckOptions opt)
```
Comment (aescande) [inline src/utils/checkFunction.cpp:463]: The TVM_DLLAPI should not be necessary here.

File: src/utils/checkFunction.cpp [hunk 6]
```diff
@@ -215,13 +374,120 @@ bool TVM_DLLAPI checkNormalAcceleration(FunctionPtr f, CheckOptions opt)
...
+       && f->isOutputEnabled(Output::Jacobian)))
+  {
+    throw std::runtime_error("The function must provide its normal acceleration, velocity and jacobian for this test.");
+  }
+
+  UpdatelessFunction uf(f);
+  return checkNormalAcceleration(uf, opt);
+}
+
 bool TVM_DLLAPI checkFunction(FunctionPtr f, CheckOptions opt)
```
Comment (aescande) [inline src/utils/checkFunction.cpp:476]: idem

File: src/utils/checkFunction.cpp (hunk 5 above)
Comment (gergondet) [reply src/utils/checkFunction.cpp]: Done in f5f60233

File: src/utils/checkFunction.cpp (hunk 6 above)
Comment (gergondet) [reply src/utils/checkFunction.cpp]: Done in f5f60233

File: src/utils/checkFunction.cpp (hunk 3 above)
Comment (gergondet) [reply src/utils/checkFunction.cpp]: Good catch! Fixed in a31726fd

File: src/utils/checkFunction.cpp (hunk 4 above)
Comment (gergondet) [reply src/utils/checkFunction.cpp]: Good catch! Fixed in a31726fd

File: include/tvm/utils/checkFunction.h (hunk 1 above)
Comment (gergondet) [reply include/tvm/utils/checkFunction.h:46]: Done in 96de425b

File: include/tvm/utils/checkFunction.h (hunk 2 above)
Comment (gergondet) [reply include/tvm/utils/checkFunction.h:61]: Agreed. Implemented in 385d57ff

**Pull request merged**